File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletTilt;

type
  TExample = class
  private
    ipcon: TIPConnection;
    al: TBrickletTilt;
  public
    procedure TiltStateCB(sender: TBrickletTilt; const tiltState: byte);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback function for tilt state callback }
procedure TExample.TiltStateCB(sender: TBrickletTilt; const tiltState: byte);
begin
  case tiltState of
    BRICKLET_TILT_TILT_STATE_CLOSED:
    begin
      WriteLn('closed');
    end;
    BRICKLET_TILT_TILT_STATE_OPEN:
	begin
      WriteLn('open');
    end;
    BRICKLET_TILT_TILT_STATE_CLOSED_VIBRATING:
	begin
      WriteLn('closed vibrating');
    end;
  end;
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  al := TBrickletTilt.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Enable tilt state callack }
  al.EnableTiltStateCallback();

  { Register tilt state callback to procedure TiltStateCB }
  al.OnTiltState := {$ifdef FPC}@{$endif}TiltStateCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Tilt Bricklet

    ' Callback subroutine for tilt state callback
    Sub TiltStateCB(ByVal sender As BrickletTilt, ByVal state As Byte)
        If state = BrickletTilt.TILT_STATE_CLOSED Then
            Console.WriteLine("Tilt State: Closed")
        Else If state = BrickletTilt.TILT_STATE_OPEN Then
            Console.WriteLine("Tilt State: Open")
        Else If state = BrickletTilt.TILT_STATE_CLOSED_VIBRATING Then
            Console.WriteLine("Tilt State: Closed Vibrating")
        End If
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim t As New BrickletTilt(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Enable tilt state callback
        t.EnableTiltStateCallback()

        ' Register tilt state callback to subroutine TiltStateCB
        AddHandler t.TiltStateCallback, AddressOf TiltStateCB

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
